cwlgen/common.py

```diff
@@ -0,0 +1,260 @@
+#  Constant(s)  ------------------------------
+
+CWL_SHEBANG = "#!/usr/bin/env cwl-runner"
+CWL_VERSIONS = ['draft-2', 'draft-3.dev1', 'draft-3.dev2', 'draft-3.dev3',
+                'draft-3.dev4', 'draft-3.dev5', 'draft-3', 'draft-4.dev1',
+                'draft-4.dev2', 'draft-4.dev3', 'v1.0.dev4', 'v1.0']
+DEF_VERSION = 'v1.0'
+
+
+#  Function(s)  ------------------------------
+
+def parse_type(param_type, requires_type=False):
+    """
+    Parses the parameter type as one of the required types:
+    :param requires_type:
+    :: https://www.commonwl.org/v1.0/CommandLineTool.html#CommandInputParameter
+
+    :param param_type: a CWL type that is _validated_
+    :type param_type: CWLType | CommandInputRecordSchema | CommandInputEnumSchema | CommandInputArraySchema | string |
+       array<CWLType | CommandInputRecordSchema | CommandInputEnumSchema | CommandInputArraySchema | string>
+    :return: CWLType | CommandInputRecordSchema | CommandInputEnumSchema | CommandInputArraySchema | string |
+       array<CWLType | CommandInputRecordSchema | CommandInputEnumSchema | CommandInputArraySchema | string>
+    """
+
+    if requires_type is False and param_type is None:
+        return None
+
+    if isinstance(param_type, str) and len(param_type) > 0:
+        # Must be CWLType
+        optional = param_type[-1] == "?"
+        if optional:
+            _LOGGER.debug("Detected {param_type} to be optional".format(param_type=param_type))
+            cwltype = param_type[:-1]
+        else:
+            cwltype = param_type
+        # cwltype = param_type[:-1] if optional else param_type
+
+        # check for arrays
+        if len(cwltype) > 2 and cwltype[-2:] == "[]":
+            array_type = CommandInputArraySchema(items=cwltype[:-2])
+            # How to make arrays optional input: https://www.biostars.org/p/233562/#234089
+            return [CwlTypes.DEF_TYPE, array_type] if optional else array_type
+
+        if cwltype not in CwlTypes.TYPES:
+            _LOGGER.warning("The type '{param_type}' is not a valid CWLType, expected one of: {types}"
+                            .format(param_type=param_type, types=", ".join(str(x) for x in CwlTypes.TYPES)))
+            _LOGGER.warning("type is set to {}.".format(CwlTypes.DEF_TYPE))
+            return CwlTypes.DEF_TYPE
+        return param_type
+
+    elif isinstance(param_type, list):
+        return [parse_type(p) for p in param_type]
+
+    elif isinstance(param_type, CommandInputArraySchema) \
+            or isinstance(param_type, CommandInputRecordSchema) \
+            or isinstance(param_type, CommandInputEnumSchema):
+        return param_type  # validate if required
+
+    if requires_type is True:
+        raise Exception("'parse_type' was required but failed to parse '{ptype}', exiting")
+
+    _LOGGER.warning("Unable to detect type of param '{param_type}'".format(param_type=param_type))
+    return CwlTypes.DEF_TYPE
+
+
+def get_type_dict(param_type):
+    """
+    Convets
```

Comment:
?

Reply:
I've clarified this to:
> Generic method to the get dict for any of the valid param_type types,
>   ie:  `CWLType | InputRecordSchema | InputEnumSchema | InputArraySchema | string | array<CWLType | InputRecordSchema | InputEnumSchema | InputArraySchema | string>`

For `type` properties such as on a [`CommandInputParameter`](https://www.commonwl.org/v1.0/CommandLineTool.html#CommandInputParameter), need to convert the accepted types into a string, list or dictionary, that's what this method does.